### PR TITLE
SnowGem Support

### PR DESCRIPTION
**Pull request by ciripel**
Please add support for SnowGem (https://snowgem.org/). We have a running dirty build here https://blockbook.snowgem.org/
Please let me know if I need to add/change something. Thank you in advance.

PS: Happy new year!

### CODE REVIEW

**Comment by martinboehm** [comment]
Hi, I successfully built the backend and blockbook packages, however when trying to install backend using `dpkg -i backend-snowgem_3.0.4-7-satoshilabs-1_amd64.deb`, I got the following error:

```
/var/lib/dpkg/info/backend-snowgem.postinst: line 19: /opt/coins/nodes/snowgem/fetch-params.sh: No such file or directory
dpkg: error processing package backend-snowgem (--install):
 subprocess installed post-installation script returned error exit status 127
```

**Comment by ciripel** [comment]
Yes that is the same thing that we did. Pls copy manually file from https://github.com/Snowgem/Data/blob/master/fetch-params.sh to /opt/coins/nodes/snowgem/
We have chosen this way to use same builds for all setups. Pls copy the file and rerun the building.

**Comment by martinboehm** [comment]
Seems I cannot install Snowgem backend on Debian 9, I get error:

```
/opt/coins/nodes/snowgem/snowgemd: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `CXXABI_1.3.11' not found (required by /opt/coins/nodes/snowgem/snowgemd)
/opt/coins/nodes/snowgem/snowgemd: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by /opt/coins/nodes/snowgem/snowgemd)
```

**Comment by ciripel** [comment]
I've fixed the files to point to a debian 9.11.0 build we have done especially for this. I've included fetch_params.sh also in the build so all the installation now go smooth. Unfortunately you will have to build again. I've tested it on a fresh debian 9.11 VM and it went smooth. Please check on your side.

**Comment by ciripel** [comment]
I have noticed some days of inactivity from your part. Hope you are good!

**Comment by martinboehm** [comment]
I would have expected something like "Thanks for fixing the merge conflict in my repo, which prevented testing of my PR" but when you are asking, thank you, I am fine :)

**Comment by martinboehm** [comment]
Hi, when I run the integration test 
`make test-integration ARGS="-v -run=TestIntegration/snowgem=main"`,
these tests fail:
```
--- FAIL: TestIntegration/snowgem=main/rpc/GetTransaction (0.00s)
--- FAIL: TestIntegration/snowgem=main/rpc/GetTransactionForMempool (0.00s)
```

**Comment by ciripel** [comment]
I just saw your answer. I will check and come back with a solution

**Comment by ciripel** [comment]
**Unit tests:** *passed*

```
$ make test
ok    blockbook/bchain/coins/snowgem  0.024s
```
**Integration tests:** *passed*

```
$ make test-integration ARGS="-v -run=TestIntegration/snowgem=main"
=== RUN   TestIntegration
=== RUN   TestIntegration/snowgem=main
=== RUN   TestIntegration/snowgem=main/rpc
=== RUN   TestIntegration/snowgem=main/rpc/GetBlock
=== RUN   TestIntegration/snowgem=main/rpc/GetBlockHash
=== RUN   TestIntegration/snowgem=main/rpc/GetTransaction
=== RUN   TestIntegration/snowgem=main/rpc/GetTransactionForMempool
=== RUN   TestIntegration/snowgem=main/rpc/MempoolSync
=== RUN   TestIntegration/snowgem=main/rpc/EstimateSmartFee
=== RUN   TestIntegration/snowgem=main/rpc/EstimateFee
=== RUN   TestIntegration/snowgem=main/rpc/GetBestBlockHash
=== RUN   TestIntegration/snowgem=main/rpc/GetBestBlockHeight
=== RUN   TestIntegration/snowgem=main/rpc/GetBlockHeader
=== RUN   TestIntegration/snowgem=main/sync
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocksParallel
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocksParallel/verifyBlockInfo
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocksParallel/verifyTransactions
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocksParallel/verifyAddresses
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocks
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocks/verifyBlockInfo
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocks/verifyTransactions
=== RUN   TestIntegration/snowgem=main/sync/ConnectBlocks/verifyAddresses
=== RUN   TestIntegration/snowgem=main/sync/HandleFork
--- PASS: TestIntegration (1.27s)
    --- PASS: TestIntegration/snowgem=main (1.27s)
        --- PASS: TestIntegration/snowgem=main/rpc (0.03s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetBlock (0.00s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetBlockHash (0.00s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetTransaction (0.00s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetTransactionForMempool (0.00s)
            --- SKIP: TestIntegration/snowgem=main/rpc/MempoolSync (0.01s)
                rpc.go:259: Skipping test, no addresses in mempool
            --- PASS: TestIntegration/snowgem=main/rpc/EstimateSmartFee (0.01s)
            --- PASS: TestIntegration/snowgem=main/rpc/EstimateFee (0.00s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetBestBlockHash (0.00s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetBestBlockHeight (0.00s)
            --- PASS: TestIntegration/snowgem=main/rpc/GetBlockHeader (0.00s)
        --- PASS: TestIntegration/snowgem=main/sync (1.16s)
            --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocksParallel (0.27s)
                --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocksParallel/verifyBlockInfo (0.00s)
                --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocksParallel/verifyTransactions (0.00s)
                --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocksParallel/verifyAddresses (0.00s)
            --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocks (0.41s)
                --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocks/verifyBlockInfo (0.00s)
                --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocks/verifyTransactions (0.00s)
                --- PASS: TestIntegration/snowgem=main/sync/ConnectBlocks/verifyAddresses (0.00s)
            --- PASS: TestIntegration/snowgem=main/sync/HandleFork (0.48s)
PASS
ok    blockbook/tests  1.403s
```

Fixed now. Sorry for the delay.

File: bchain/coins/blockchain.go [hunk 1]
```diff
@@ -107,7 +108,7 @@ func init() {
 	BlockChainFactories["CPUchain"] = cpuchain.NewCPUchainRPC
 	BlockChainFactories["Unobtanium"] = unobtanium.NewUnobtaniumRPC
 	BlockChainFactories["DeepOnion"] = deeponion.NewDeepOnionRPC
-	BlockChainFactories["SnowGem"] = zec.NewZCashRPC
```
Comment (IKE471) [inline bchain/coins/blockchain.go]: shows red

File: bchain/coins/snowgem/snowgemparser_test.go [hunk 2]
```diff
@@ -18,8 +18,8 @@ import (
...
...= "0a20241803e368d7459f31286a155191ee386896d366d57c19d8e67a8f040d6ff71f12ab100400008085202f890119950c49d69b37d5f4fbb390d852387559e6a6d3fce9f390a409e4acf3f06381020000006a4730440220452aedf599e575598eb36d27ed98a6d388efda6e9be2bab96f16d0644e7df3060220669f4f3a4976ed73fa3ca9ecaad84dcf6ec35099c3bad631499985ea6a378d19012102ed9fb7fb61ec514be890ab45a925d554ff12050f099514251d5ebe904accc93ffeffffff02d3d0a146000000001976a9141a78c04d87f553545ba225b7bc7a271731f659d688ac7c54ae02000000001976a914b86f4b063545ebc2e80522a59d2dd206b707401b88aca68d0e00c58d0e00000000000000000000000018a0f1c9d505200028b0eb113299010a0012208163f0f3ace409a490f3e9fcd3a6e659753852d890b3fbf4d5379bd6490c95191800226b4730440220452aedf599e575598eb36d27ed98a6d388efda6e9be2bab96f16d0644e7df3060220669f4f3a4976ed73fa3ca9ecaad84dcf6ec35099c3bad631499985ea6a378d19012102ed9fb7fb61ec514be890ab45a925d554ff12050f099514251d5ebe904accc93f28ffffffff0f3a490a05043c1aec8e10001a1976a9141a78c04d87f553545ba225b7bc7a271731f659d688ac22237431566d4854547770457477766f6a786f644e32435351714c596931687a59336341713a470a0398968010011a1976a914b86f4b063545ebc2e80522a59d2dd206b707401b88ac222374315934794c31344143486141626a656d6b647057376e594e48576e763179516244414000"
```
Comment (IKE471) [inline bchain/coins/snowgem/snowgemparser_test.go]: 21,22 shows red

File: tests/rpc/testdata/snowgem.json [hunk 3]
```diff
@@ -11,7 +11,7 @@
    ],
    "txDetails":{
       "b1ab0b7bc4f30dea352ba6b641777d0ed6157027625aacb36da1c5eedc28e157":{
-         "hex":"0400008085202f8901d79a4bae6a56937ddbe26fa44ad8eedc389a415fcbbf75a46ef1500cd0837e5d000000006a47304402207d7268375db5a1544368fa9a3c4e77244bc36c3ecf18b21a409f267f1436aac802202e0d7f4abafab362607d70c88a35cc50bfb4f5643d9f8d5208f5db0862946dd2012103af4c37554efb4420c42edc2f7a84e3e687dc050a832afb6ac5bcc12cfb903fbffeffffff0584f1f81f000000001976a914b38cf949ff45875224793172f8333c43e657e3e888aca61ef369000000001976a914559f99e0b3f83666cb0ef0b55c1c367b68e1d65988ac20d55023000000001976a9147a7b884e38d58bbfffbc0a0f0bee44336ca1e8a88ac205c5071000000001976a9149d0da224652d49091082f0b1c2e2b713cbceb47a88ac20d55023000000001976a91406a5e5f229d7bfd6b979747c92e6e1f6dcc95c3488ac74010e0093010e000000000000000000000000",
```
Comment (IKE471) [inline tests/rpc/testdata/snowgem.json]: shows red

File: tests/rpc/testdata/snowgem.json [hunk 4]
```diff
@@ -147,4 +147,4 @@
          ]
       }
    }
-}
```
Comment (IKE471) [inline tests/rpc/testdata/snowgem.json]: shows red

File: tests/rpc/testdata/snowgem.json (hunk 3 above)
Comment (ciripel) [reply tests/rpc/testdata/snowgem.json]: u are reviewing initial implementation not after the fix

**Pull request merged**